src/main/java/bio/terra/service/auth/iam/IamService.java

```diff
@@ -148,6 +150,36 @@ public Map<UUID, Set<IamRole>> listAuthorizedResources(
     return callProvider(() -> iamProvider.listAuthorizedResources(userReq, iamResourceType));
   }
 
+  /**
+   * Note: calling this method will trigger a call to SAM API. No cache backs it. If calling,
+   * consider whether new usage patterns necessitate refactoring to add a cache.
+   *
+   * @param userReq authenticated user
+   * @param iamResourceType resource type; e.g. dataset
+   * @param resourceId UUID identifying a resource
+   * @param actions required
+   * @throws IamForbiddenException if the user is missing any `actions` on the resource
+   */
+  public void verifyAuthorizations(
+      AuthenticatedUserRequest userReq,
+      IamResourceType iamResourceType,
+      String resourceId,
+      List<IamAction> actions)
+      throws IamForbiddenException {
+    String userEmail = userReq.getEmail();
+    List<IamAction> availableActions =
+        callProvider(() -> iamProvider.listActions(userReq, iamResourceType, resourceId)).stream()
+            .map(IamAction::fromValue)
+            .collect(Collectors.toList());
+
+    List<IamAction> unavailableActions = new ArrayList<>(actions);
+    unavailableActions.removeAll(availableActions);
+    if (!unavailableActions.isEmpty()) {
+      throw new IamForbiddenException(
+          "User '" + userEmail + "' does not have required action(s): " + unavailableActions);
```

Comment:
Another benefit of this change: if the user is missing multiple actions, all of the missing actions will be logged in the thrown exception.  Previously we'd throw on the first missing.